NativeRegExp: Implement named capture groups

balajirrao opened the pull request:
**Note: This PR also includes commits from the lookbehind [PR](https://github.com/mozilla/rhino/pull/1843).**

This adds partial support for named capture groups in NativeRegExp. This also adds support for [duplicate-named-capturing-groups](https://github.com/tc39/proposal-duplicate-named-capturing-groups).

What is not implemented yet:
- capture group names that consist of unicode escape sequences.
- String.prototype.{replace, replaceAll}
 
Many new test262 test cases pass and there are no regressions. The ones that are failing are due to
- 'u' regexp flag not supported
- objects will null prototype do not work correctly with respect to accessing `__proto__` (a fix is being worked on)
- existing bugs in String.prototype.{replace, replaceAll, split}
- no support for classes

rbri commented:
Great

rbri commented:
@balajirrao thanks for all the work on the regexp support.

Since ages HtmlUnit uses the rhino regexp more as a backup and tries to translate the js regexp into java regexp. Based on your work i like to remove that stuff and stay with the rhino impl.
But while updating i found another thing - web pages also using the js regexp support for validating input fields. This means i have to use the regexp stuff without a js environment (more or less). Or long story short...

What do you think about separating the regexp stuff from the js stuff meaning moving most of the static methods and all the helper classes into a separate (static) class?
Have done some prototyping and it seems to work well. Only some methods still use a Context parameter. Currently i think about options for that.
This might also help a bit to write more basic unit tests.

So what do you think?

balajirrao commented:
@rbri thanks for your comment!

What you are saying sounds very reasonable. I have a couple of regexp improvements on the plate before I can tackle this. In case you have your prototype on a branch, I'd be interested to take a look!

rbri commented:
@ balajirrao

> I have a couple of regexp improvements in the pipeline before I can tackle this.
Yes please....

> In case you have your prototype on a branch, I'd be interested to take a look!
I did something in this direction but i was not really happy with my solution.

Will work on that maybe during the next days and prepare a draft pr to have some base to discuss.

rbri commented:
@balajirrao 

As starting point.... 56567bc71c4150285a44b3e60857fb05ad7dd696

balajirrao commented:
@gbrail I figured you must have started reviewing this and hence I've pushed some changes I'd like to be included in the PR, in a separate commit. The changes improves how named capture groups are parsed and has some refactoring to pave the way for unicode support (which I'm working on ATM). The commit message has some info. 

Some background:
While working on unicode regexp support , I discovered https://tc39.es/ecma262/#sec-regular-expressions-patterns, which I also figured is what Rhino supports. This led me to extract the parser for the CharacterClassEscape and CharacterEscape non-terminals,  which can be used in parsing both the AtomEscape and the ClassEscape non-terminals (the code to do this was duplicated in 3 different places in the code before).

andreabergia commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -888,150 +884,46 @@ private static boolean parseAlternative(CompilerState state) {
+
+        for (RENode node : classContents.escapeNodes) {
+            if (node.op != REOP_FLAT) {
+                target.bmsize = 65536;
```
Can we use a named constant instead of a magic number?

andreabergia commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -1175,7 +1067,285 @@ private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend
+    private static ClassContents parseClassContents(CompilerState state, ParserParameters params) {
+        char[] src = state.cpbegin;
+        char rangeStart = 0;
+        ;
```
Spurious

andreabergia commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -1175,7 +1067,285 @@ private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend
+        doFlat(state, c);
+    }
+
+    static class ClassContents {
+        boolean sense;
+        ArrayList<Character> chars;
+        ArrayList<Character> ranges; // ranges stored as (start1, end1, start2, end2, ...)
+        ArrayList<RENode> escapeNodes;
+
+        ClassContents() {
+            chars = new ArrayList<>();
+            ranges = new ArrayList<>();
+            escapeNodes = new ArrayList<>();
+            sense = true;
+        }
+    }
```
Why not do this in the field declaration?

```suggestion
    static class ClassContents {
        boolean sense = true;
        ArrayList<Character> chars = new ArrayList<>();
        ArrayList<Character> ranges = new ArrayList<(); // ranges stored as (start1, end1, start2, end2, ...)
        ArrayList<RENode> escapeNodes = new ArrayList<>();
    }
```

gbrail reviewed (changes requested):
Thanks -- great progress, it looks like we have some minor comments from me and also Andrea so please take a look.

Also, I didn't catch this in a previous, but I see that the class now has a "debug" flag. We actually have a whole mechanism for those that lets you easily set it at runtime -- take a look at what's done in the class below and please do it the same way:

https://github.com/mozilla/rhino/blob/f8ce76a38b172521a9e49050ebfb83de0bb5e073/rhino/src/main/java/org/mozilla/javascript/Token.java#L29

Finally, please take a look at test coverage for the new code -- I didn't look yet. Thanks!

balajirrao commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -78,7 +82,8 @@ public class NativeRegExp extends IdScriptableObject {
     private static final byte REOP_NCLASS = 23; /* negated character class with index */
-    private static final byte REOP_SIMPLE_END = 23; /* end of 'simple opcodes' */
+    private static final byte REOP_NAMED_BACKREF = 24; /* named back-reference */
+    private static final byte REOP_SIMPLE_END = 24; /* end of 'simple opcodes' */
```
It is fine for them to have the same values. REOP_SIMPLE_END is not an operator, but used as a marker to determine if an op is simple using `op <= REOP_SIMPLE_END`. 

The `+1` pattern is good. I'll switch them in a next PR, right after this one :)

balajirrao commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -536,6 +550,61 @@ private static void prettyPrintRE(RECompiled regexp) {
+                    String name =
+                            new String(
+                                    src, node.captureGroupNameIndex, node.captureGroupNameLength);
+                    ArrayList<Integer> entry = new ArrayList<>(1);
```
Not quite. I create an arraylist of size 1 optimistically assuming that there are no duplicates. However, if we find duplicates (allowed in the case of disjunctions), the list will have more elements than one. I think this deserves a comment in the code.

balajirrao commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -1061,7 +1035,317 @@ private static RENode reverseNodeList(RENode head) {
-    private static boolean parseTerm(CompilerState state) {
+    private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend) {
+        // TODO: support capture group names consisting of unicode escape sequences
+        String srcString = new String(src);
```
Good catch - I'm not very familiar with the unicode api in Character and String. As soon I as I got something working I was happy to move on 😅. Will fix.

balajirrao commented:
@gbrail thank you for taking a look! I've addressed the comments. I also took at look at the test coverage and added couple tests for a couple of edge cases.

I'll fix the debug flag in a next PR - it has been around before I started with Rhino.

gbrail commented:
Can you please just comment that constant with a duplicate value? Other
people will look at it and assume it's a mistake.

On Mon, Mar 24, 2025 at 8:43 AM Balaji Rao ***@***.***> wrote:

> @balajirrao <https://github.com/balajirrao> requested your review on:
> #1847 <https://github.com/mozilla/rhino/pull/1847> NativeRegExp:
> Implement named capture groups.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/rhino/pull/1847#event-16958412902>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7I2424HDONCZ62HGKEYT2WAR2FAVCNFSM6AAAAABYL6GREOVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJWHE2TQNBRGI4TAMQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

gbrail commented:
Thanks for all your work on this!

balajirrao commented:
@gbrail thank you for your review and for merging it! I think I'm finally getting the hang of the regexp parser.
